### PR TITLE
[22.05] Ensure guid is hexadecimal value

### PR DESCRIPTION
**Pull request by mvdbeek**
Fixes https://github.com/galaxyproject/galaxy/issues/14560:

```
File '/gpfs1/data/galaxy_server/galaxy-dev/lib/galaxy/web/framework/middleware/error.py', line 165 in __call__
  app_iter = self.application(environ, sr_checker)
File '/gpfs1/data/galaxy_server/galaxy-dev/.venv/lib/python3.8/site-packages/paste/recursive.py', line 85 in __call__
  return self.application(environ, start_response)
File '/gpfs1/data/galaxy_server/galaxy-dev/.venv/lib/python3.8/site-packages/paste/httpexceptions.py', line 640 in __call__
  return self.application(environ, start_response)
File '/gpfs1/data/galaxy_server/galaxy-dev/lib/galaxy/web/framework/base.py', line 166 in __call__
  return self.handle_request(environ, start_response)
File '/gpfs1/data/galaxy_server/galaxy-dev/lib/galaxy/web/framework/base.py', line 222 in handle_request
  trans = self.transaction_factory(environ)
File '/gpfs1/data/galaxy_server/galaxy-dev/lib/galaxy/webapps/base/webapp.py', line 101 in <lambda>
  self.set_transaction_factory(lambda e: self.transaction_chooser(e, galaxy_app, session_cookie))
File '/gpfs1/data/galaxy_server/galaxy-dev/lib/galaxy/webapps/base/webapp.py', line 190 in transaction_chooser
  return GalaxyWebTransaction(environ, galaxy_app, self, session_cookie)
File '/gpfs1/data/galaxy_server/galaxy-dev/lib/galaxy/webapps/base/webapp.py', line 311 in __init__
  self._ensure_valid_session(session_cookie)
File '/gpfs1/data/galaxy_server/galaxy-dev/lib/galaxy/webapps/base/webapp.py', line 557 in _ensure_valid_session
  galaxy_session = self.session_manager.get_session_from_session_key(session_key=session_key)
File '/gpfs1/data/galaxy_server/galaxy-dev/lib/galaxy/managers/session.py', line 26 in get_session_from_session_key
  self.sa_session.query(self.model.GalaxySession)
File '/gpfs1/data/galaxy_server/galaxy-dev/.venv/lib/python3.8/site-packages/sqlalchemy/orm/query.py', line 2819 in first
  return self.limit(1)._iter().first()
File '/gpfs1/data/galaxy_server/galaxy-dev/.venv/lib/python3.8/site-packages/sqlalchemy/orm/query.py', line 2903 in _iter
  result = self.session.execute(
File '/gpfs1/data/galaxy_server/galaxy-dev/.venv/lib/python3.8/site-packages/sqlalchemy/orm/session.py', line 1712 in execute
  result = conn._execute_20(statement, params or {}, execution_options)
File '/gpfs1/data/galaxy_server/galaxy-dev/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py', line 1631 in _execute_20
  return meth(self, args_10style, kwargs_10style, execution_options)
File '/gpfs1/data/galaxy_server/galaxy-dev/.venv/lib/python3.8/site-packages/sqlalchemy/sql/elements.py', line 325 in _execute_on_connection
  return connection._execute_clauseelement(
File '/gpfs1/data/galaxy_server/galaxy-dev/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py', line 1498 in _execute_clauseelement
  ret = self._execute_context(
File '/gpfs1/data/galaxy_server/galaxy-dev/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py', line 1862 in _execute_context
  self._handle_dbapi_exception(
File '/gpfs1/data/galaxy_server/galaxy-dev/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py', line 2047 in _handle_dbapi_exception
  util.raise_(exc_info[1], with_traceback=exc_info[2])
File '/gpfs1/data/galaxy_server/galaxy-dev/.venv/lib/python3.8/site-packages/sqlalchemy/util/compat.py', line 208 in raise_
  raise exception
File '/gpfs1/data/galaxy_server/galaxy-dev/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py', line 1819 in _execute_context
  self.dialect.do_execute(
File '/gpfs1/data/galaxy_server/galaxy-dev/.venv/lib/python3.8/site-packages/sqlalchemy/engine/default.py', line 732 in do_execute
  cursor.execute(statement, parameters)
ValueError: A string literal cannot contain NUL (0x00) characters.
```

This got logged because https://github.com/galaxyproject/galaxy/commit/29bba7316a1ca22e0f7575f90d7b2b2b50e23d0c narrowed the exception catching, but we didn't anticipate the ValueError raised by sqlalchemy. Ensuring the decoded session key is a hexadecimal value seems like a clean fix.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Review by davelopez — APPROVED** [review]

File: test/unit/data/security/test_id_encode_decode.py [hunk 1]
```diff
@@ -62,6 +62,16 @@ def test_maximum_length_handling_nonascii():
     assert e11 != e12
 
 
+def test_unicode_null_decoding():
+    encoded_id = test_helper_1.encode_id(1)
+    try:
```
Comment (nsoranzo) [inline test/unit/data/security/test_id_encode_decode.py:68]:
```suggestion
    threw_exception = False
    try:
```

File: lib/galaxy/security/idencoding.py [hunk 2]
```diff
@@ -106,7 +106,10 @@ def decode_guid(self, session_key):
         # Session keys are strings
```
Comment (nsoranzo) [inline lib/galaxy/security/idencoding.py:109]: Maybe a good time to add type annotation to this method?

```python
    def decode_guid(self, session_key: Union[bytes, str]) -> str:

Comment (mvdbeek) [reply lib/galaxy/security/idencoding.py:109]: Then we have to type unicodify as well, which brings 28 other issues.

Comment (mvdbeek) [reply lib/galaxy/security/idencoding.py:109]: With type overloads it's not actually that bad, I'll push a commit

Comment (nsoranzo) [reply lib/galaxy/security/idencoding.py:109]: Ah, sorry, I didn't want to open a Pandora's box!

File: lib/galaxy/util/__init__.py [hunk 3]
```diff
@@ -32,6 +32,13 @@
 from hashlib import md5
 from os.path import relpath
 from pathlib import Path
+from typing import (
+    Any,
+    Literal,
```
Comment (nsoranzo) [inline lib/galaxy/util/__init__.py]: This needs to come from typing_extensions for Python 3.7 compatibility: `from typing_extensions import Literal`

File: lib/galaxy/util/__init__.py [hunk 4]
```diff
@@ -32,6 +32,12 @@
 from hashlib import md5
 from os.path import relpath
 from pathlib import Path
+from typing import (
+    Any,
+    Optional,
+    overload,
+    Union,
```
Comment (nsoranzo) [inline lib/galaxy/util/__init__.py]:
```suggestion
```

File: lib/galaxy/util/__init__.py [hunk 5]
```diff
@@ -1083,7 +1090,35 @@ def roundify(amount, sfs=2):
         return amount[0:sfs] + "0" * (len(amount) - sfs)
 
 
-def unicodify(value, encoding=DEFAULT_ENCODING, error="replace", strip_null=False, log_exception=True):
+@overload
+def unicodify(
```
Comment (nsoranzo) [inline lib/galaxy/util/__init__.py]:
```suggestion
def unicodify(  # type: ignore[misc]
```

File: lib/galaxy/util/__init__.py [hunk 6]
```diff
@@ -1083,7 +1090,35 @@ def roundify(amount, sfs=2):
...
+    error: str = "replace",
+    strip_null: bool = False,
+    log_exception: bool = True,
+) -> None:
+    ...
+
+
+@overload
+def unicodify(
+    value: Union[str, bytes, int, float, bool, Exception],  # Anything but None, but this seems fine for now
```
Comment (nsoranzo) [inline lib/galaxy/util/__init__.py]:
```suggestion
    value: Any,
```

**Review by nsoranzo — APPROVED** [review]

**Pull request merged**